src/frames.ts

```diff
@@ -889,84 +888,78 @@ export class Frame {
-  private _contextData: ContextData;
-  private _task: Task;
-  private _runCount: number;
-  private _resolve: (result: js.JSHandle) => void = () => {};
+class RerunnableTask<T> {
+  readonly promise: Promise<types.SmartHandle<T>>;
+  terminate: (reason: Error) => void = () => {};
+  private _task: SchedulableTask<T>;
+  private _resolve: (result: types.SmartHandle<T>) => void = () => {};
   private _reject: (reason: Error) => void = () => {};
-  private _timeoutTimer?: NodeJS.Timer;
-  private _terminated = false;
+  private _terminatedPromise: Promise<Error>;
 
-  constructor(data: ContextData, task: Task, deadline: number, title?: string) {
-    this._contextData = data;
+  constructor(data: ContextData, task: SchedulableTask<T>, deadline: number, title?: string) {
     this._task = task;
-    this._runCount = 0;
-    this.promise = new Promise<js.JSHandle>((resolve, reject) => {
-      this._resolve = resolve;
-      this._reject = reject;
-    });
+    data.rerunnableTasks.add(this);
+
     // Since page navigation requires us to re-install the pageScript, we should track
     // timeout on our end.
     const timeoutError = new TimeoutError(`waiting for ${title || 'function'} failed: timeout exceeded. Re-run with the DEBUG=pw:input env variable to see the debug log.`);
-    this._timeoutTimer = setTimeout(() => this.terminate(timeoutError), helper.timeUntilDeadline(deadline));
-  }
+    let timeoutTimer: NodeJS.Timer | undefined;
+    this._terminatedPromise = new Promise(resolve => {
+      timeoutTimer = setTimeout(() => resolve(timeoutError), helper.timeUntilDeadline(deadline));
+      this.terminate = resolve;
+    });
+
+    // This promise is either resolved with the task result, or rejected with a meaningful
+    // evaluation error.
+    const resultPromise = new Promise<types.SmartHandle<T>>((resolve, reject) => {
+      this._resolve = resolve;
+      this._reject = reject;
+    });
+    const failPromise = this._terminatedPromise.then(error => Promise.reject(error));
 
-  terminate(error: Error) {
-    this._terminated = true;
-    this._reject(error);
-    this._doCleanup();
+    this.promise = Promise.race([resultPromise, failPromise]).finally(() => {
+      if (timeoutTimer)
+        clearTimeout(timeoutTimer);
+      data.rerunnableTasks.delete(this);
+    });
   }
 
   async rerun(context: dom.FrameExecutionContext) {
-    const runCount = ++this._runCount;
-    let success: js.JSHandle | null = null;
-    let error = null;
+    let progress: js.JSHandle<types.InjectedScriptProgress<T>> | null = null;
+
+    // On timeout or error, cancel current progress.
+    const cleanupProgress = () => {
+      if (!progress)
+        return;
+      const copy = progress;
+      progress = null;
+      copy.evaluate(p => p.cancel()).catch(e => {}).then(() => copy.dispose());
+    };
+    this._terminatedPromise.then(cleanupProgress);
+
     try {
-      success = await this._task(context);
+      progress = await this._task(context);
```

Comment:
This progress conceptually clashes with the one from the previous patch. Can we align them?